File: IPWorks Samples/Atom Feed Reader/java/atom.java
/*
 * IPWorks 2022 Java Edition - Sample Project
 *
 * This sample project demonstrates the usage of IPWorks in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/ipworks
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */

import java.io.*;
import ipworks.*;

public class atom extends ConsoleDemo {
  
	public static void main(String[] args) {
		
		Atom atom1 = new Atom();
		
		try {
			atom1.addAtomEventListener(new DefaultAtomEventListener(){
				public void SSLServerAuthentication(AtomSSLServerAuthenticationEvent e) {
					e.accept = true; //this will trust all certificates and it is not recommended for production use
				}
			});						
			System.out.println("Fetching feed...");      
			atom1.setFollowRedirects(Atom.frAlways);
			atom1.getFeed("https://news.google.com/news?ned=us&topic=h&output=atom");
			System.out.println("Atom Channel: " + atom1.getChannel().getTitle());
			String subtitle = atom1.getChannel().getSubtitle();
			if (subtitle != null && !subtitle.equals("")) {
				System.out.print(" - " + subtitle);
			}
			for (int i = 1; i <= atom1.getEntries().size(); i++) {
				AtomEntry atomEntry = atom1.getEntries().item(i-1);
			    System.out.println(i + ". " + atomEntry.getTitle());
			}

			while (true) {
				System.out.println("\nThere are " + atom1.getEntries().size() + " entries in this feed.");
				String command = prompt("Atom Entry (Q to quit)");
				if (command.equals("Q") || command.equals("q")) {
					break;
				}

				int index = Integer.parseInt(command);
				AtomEntry atomEntry = atom1.getEntries().item(index-1);

				System.out.println("\n" + atomEntry.getContent());
				System.out.println("\nFind full article at " + atomEntry.getLinkHref());
		
			}
			
		} catch (Exception e) {
			displayError(e);
		}
	}  
}

class ConsoleDemo {
  private static BufferedReader bf = new BufferedReader(new InputStreamReader(System.in));

  static String input() {
    try {
      return bf.readLine();
    } catch (IOException ioe) {
      return "";
    }
  }
  static char read() {
    return input().charAt(0);
  }

  static String prompt(String label) {
    return prompt(label, ":");
  }
  static String prompt(String label, String punctuation) {
    System.out.print(label + punctuation + " ");
    return input();
  }

  static String prompt(String label, String punctuation, String defaultVal)
  {
	System.out.print(label + " [" + defaultVal + "] " + punctuation + " ");
	String response = input();
	if(response.equals(""))
		return defaultVal;
	else
		return response;
  }

  static char ask(String label) {
    return ask(label, "?");
  }
  static char ask(String label, String punctuation) {
    return ask(label, punctuation, "(y/n)");
  }
  static char ask(String label, String punctuation, String answers) {
    System.out.print(label + punctuation + " " + answers + " ");
    return Character.toLowerCase(read());
  }

  static void displayError(Exception e) {
    System.out.print("Error");
    if (e instanceof IPWorksException) {
      System.out.print(" (" + ((IPWorksException) e).getCode() + ")");
    }
    System.out.println(": " + e.getMessage());
    e.printStackTrace();
  }
}

File: IPWorks Samples/Atom Feed Reader/cpp/atom.cpp
/*
 * IPWorks 2022 C++ Edition - Sample Project
 *
 * This sample project demonstrates the usage of IPWorks in a 
 * simple, straightforward way. It is not intended to be a complete 
 * application. Error handling and other checks are simplified for clarity.
 *
 * www.nsoftware.com/ipworks
 *
 * This code is subject to the terms and conditions specified in the 
 * corresponding product license agreement which outlines the authorized 
 * usage and restrictions.
 */

#include <stdio.h>
#include <string.h>
#include <stdlib.h>
#include "../../include/ipworks.h"
#define LINE_LEN 100
class MyAtom : public Atom
{
public:
	
	virtual int FireSSLServerAuthentication(AtomSSLServerAuthenticationEventParams *e)
	{
		if (e->Accept) return 0;
		printf("Server provided the following certificate:\nIssuer: %s\nSubject: %s\n",
			e->CertIssuer, e->CertSubject);
		printf("The following problems have been determined for this certificate: %s\n", e->Status);
		printf("Would you like to continue anyways? [y/n] ");
		if (getchar() == 'y') e->Accept = true;
		else exit(0);
		return 0;
	}
};

int main(int argc, char **argv)
{
	char command[5]; //users input command
	bool quit = false;

	MyAtom atom1;
	atom1.SetFollowRedirects(1); // frAlways
	atom1.GetFeed("https://news.google.com/news?ned=us&topic=h&output=atom");

	while (!quit)
	{
		printf("Atom Channel :%s\n%s\n", atom1.GetChannelTitle(),atom1.GetChannelSubtitle());
		for (int i = 0; i< atom1.GetEntryCount(); i++)
		{
			printf("%i. %s\n", i+1, atom1.GetEntryTitle(i));
		}
		printf("Q Quit,news item number>");
		fgets(command,5,stdin);
		command[strlen(command)-1] = '\0';

		if ( ! strcmp(command, "Q") )
		{
			quit = true;
		}
		else
		{
			int selectedindex = atoi(command)-1;
			printf("\n\n[%d] %s\n\n",selectedindex,atom1.GetEntryContent(selectedindex));
			printf("Find full article at %s\n\n", atom1.GetEntryLinkHref(selectedindex));
		}

	}

	fprintf(stderr, "\npress <return> to continue...\n");
	getchar();
	return 0;
}








